ykllvmwrap/src/jitmodbuilder.cc

```diff
@@ -623,6 +623,16 @@ class JITModBuilder {
     }
   }
 
+  // Determines if the LLVM values `V1` and `V2` are instructions defined
+  // within the same LLVM `BasicBlock`.
+  bool isDefinedInSameBlock(Value *V1, Value *V2) {
+    if (!isa<Instruction>(V1) || !isa<Instruction>(V2))
```

Comment:
I wonder if we should say that this function only accepts `Instruction`s and this can then be an `assert`?

Reply:
It's designed this way so that the caller can determine if two values are instructions in the same block without having to determine if either value is a non-instruction type (e.g. a global or a constant) at each call site.

Reply:
I understand what it *can* do: I'm suggesting it might be doing too much and that it might not be useful to allow the user to check if non-instructions are in the same module (on the basis that we won't use that use case and thus if it ever happens it's a bug that won't be caught by `return false`). I'm happy to be wrong and for you to tell me that it does need to support non-instructions though!

Reply:
After experimentation, what I don't like about pulling the instruction check logic outside of the function is that you end up duplicating the map insertion and that junks up the call site.

```diff
diff --git a/ykllvmwrap/src/jitmodbuilder.cc b/ykllvmwrap/src/jitmodbuilder.cc
index 3fd3661..d9114ed 100644
--- a/ykllvmwrap/src/jitmodbuilder.cc
+++ b/ykllvmwrap/src/jitmodbuilder.cc
@@ -625,11 +625,7 @@ class JITModBuilder {
 
   // Determines if the LLVM values `V1` and `V2` are instructions defined
   // within the same LLVM `BasicBlock`.
-  bool isDefinedInSameBlock(Value *V1, Value *V2) {
-    if (!isa<Instruction>(V1) || !isa<Instruction>(V2))
-      return false;
-    Instruction *I1 = cast<Instruction>(V1);
-    Instruction *I2 = cast<Instruction>(V2);
+  bool isDefinedInSameBlock(Instruction *I1, Instruction *I2) {
     return I1->getParent() == I2->getParent();
   }
 
@@ -737,10 +733,13 @@ public:
       //  We need an entry in this map for any live variable that isn't defined
       //  by a PHI node at the top of the block cotaining the call to the
       //  control point.
-      if (!(isa<PHINode>(InsertedOperand) &&
-            isDefinedInSameBlock(InsertedOperand, IVI))) {
-        InsertValueMap[*IVI->idx_begin()] = IVI->getInsertedValueOperand();
-      }
+                       if (isa<Instruction>(InsertedOperand)) {
+                               Instruction *InsertedOperandInst = cast<Instruction>(InsertedOperand);
+                               if (!(isDefinedInSameBlock(InsertedOperandInst, IVI) && isa<PHINode>(InsertedOperandInst)))
+                                       InsertValueMap[*IVI->idx_begin()] = IVI->getInsertedValueOperand();
+                       } else
+                               InsertValueMap[*IVI->idx_begin()] = IVI->getInsertedValueOperand();
+
       OutS = IVI->getAggregateOperand();
     }
```

Reply:
I don't think I've got my point across very well -- sorry!

What I'm asking is: do we *ever* ask whether a non-`Instruction` is part of a module? My guess based on my understanding of the code is "no".

Reply:
Currently we don't (based on current test coverage), but I don't see why we couldn't.

`InsertedOperandInst` could easily be a global or a constant. Neither of which are an `Instruction` that can be parented by a `BasicBlock` according to the LLVM type system.

Reply:
I understand it *can* be those, but based on the current code I think `return false` will lead to bugs which `assert(isInstruction(value))`.

Reply:
I don't follow. How could it lead to bugs? Can you give an example?

Reply:
Isn't a `PHINode` always an `Instruction`?

Reply:
Yes, but it's the other operand that I'm not sure about.

Reply:
@ltratt try 0b8e03d4d3bc69885366423b519e15b9ba38f244.

I was hoping to prove you wrong with tests, but failed. Thus I've changed the name of the function and changed the `return false` case to an assertion.

Nothing is set in stone, and we can change this if a test later proves us wrong.

Reply:
What I meant was that we only ever call `isDefinedInSameBlock` on `PHINode`s, so an assert would be fine there.

Reply:
Works for me!

Reply:
> What I meant was that we only ever call isDefinedInSameBlock on PHINodes, so an assert would be fine there.

Unless I'm mistaken, the operands we pass to `areInstsDefinedInSameBlock()` are: `IVI->getInsertedOperand()` which might be a PHI; and `IVI` which is never a PHI.

So we call the function on "maybe PHI", but (we think) "always instruction" values.